Datahub: Add quick access menu

tkohr opened the pull request:
### Description

This PR introduces a new quick access "navigation bar" that scrolls down to the section of a dataset, service or reuse page. 

Note: So far the nav bar has been tested on the dataset page. Once the service and reuse page are present/merged we must ensure that the anchor ids in the page and in the nav bar "config" are the same.

<!--
Describe here the changes brought by this PR. Do not forget to link any relevant issue or discussion to help people review your work!
-->

<!--
Describe here any changes to the project architecture: adding/removing modules or libraries, changing dependencies between libraries and apps, changes to external NPM dependencies...
-->

### Screenshots

Desktop
![image](https://github.com/user-attachments/assets/7739bf23-d60d-4461-82cc-911d117c0d08)

Mobile
![image](https://github.com/user-attachments/assets/e552585b-3627-4800-828d-a9629de87ba3)


<!--
If the changes incur visual changes, please include screenshots or an animated screen capture.
-->

### Quality Assurance Checklist

- [ ] Commit history is devoid of any _merge commits_ and readable to facilitate reviews
- [ ] If **new logic** ⚙️ is introduced: unit tests were added
- [ ] If **new user stories** 🤏 are introduced: E2E tests were added
- [ ] If **new UI components** 🕹️ are introduced: corresponding stories in Storybook were created
- [ ] If **breaking changes** 🪚 are introduced: add the `breaking change` label
- [ ] If **bugs** 🐞 are fixed: add the `backport <release branch>` label
- [ ] The [documentation website](docs) 📚 has received the love it deserves

<!--
Please only check items relevant to your contribution. Thank you very much for your time and efforts!
-->

### How to test

Go to a dataset page.

To display the language switcher and test the outcome in the same colors as in the mockups, the following values can be set in the `default.toml`:

```
languages = ['en', 'fr', 'de']

primary_color = "#90C149"
secondary_color = "#F18345"
```

<!--
Describe here the steps to confirm that the changes work as they should.
-->

---

<!--
Please give credit to the sponsor of this work if possible.
-->

<!-- **This work is sponsored by [Organization ABC](xx)**. -->

jahow commented on tailwind.base.config.js:
```diff
@@ -69,6 +69,7 @@ module.exports = {
       },
       boxShadow: {
         'xl-hover': '0 20px 25px -5px rgb(0 0 0 / 0.2)',
+        navbar: '0px 0px 5.1px 0px var(--color-primary-darkest)',
```
This is not appropriate I think; why not specify the box shadow in the HTML?

tkohr replied:
Ok, wasn't sure this was possible. What's the syntax for this?

jahow replied:
not sure but if needed, you can just use CSS instead of tailwind classes

tkohr replied:
Hum, ok I thought it would be preferable to have a custom tailwind class than CSS in our components. Does that mean that in the long run we should get rid of all the custom classes in the tailwind config?

jahow replied:
In my opinion yes, global classes are technical debt unless they adress standard components like cards, buttons, icons, inputs...

jahow replied:
Using tailwind classes or css in the component is honestly almost the same, sure we try to have as few CSS as possible in the project but it's preferable to have local CSS than global classes

tkohr replied:
Ok I used tailwind interpolation

jahow commented on tailwind.base.css:
```diff
@@ -157,6 +161,11 @@
   }
 
+  /* TODO: rework centering in gn-ui-card-icon and maybe merge both classes into one */
+  .gn-ui-nav-bar-icon {
+    @apply gn-ui-card-icon justify-center items-center h-8 min-w-10 hover:bg-primary;
+  }
```
this shouldn't be a new global class; global classes are for very generic elements like buttons, cards, input fields etc.

tkohr replied:
Ok I'll try to use the `gn-ui-button` for this. I guess the `favorite-star` component can be wrapped into it.

tkohr replied:
I refactored the toggle-star component to use the `gn-ui-button` and thus become "themable"

jahow commented on libs/feature/search/src/lib/favorites/favorite-star/favorite-star.component.html:
```diff
@@ -10,10 +10,4 @@
-    class="align-text-top ml-1.5"
-    data-test="favorite-count"
-    *ngIf="!displayCount"
-    translate="datahub.record.addToFavorites"
```
Maybe this translation key could be kept as a tooltip? It is a useful indication IMO (and probably required in terms of accessibility)

tkohr replied:
It was actually not displayed anymore before this PR, so I've left it out for now. Displaying it as a simple tooltip could create a conflict with the "login message tooltip", displaying it in a tippy tooltip might be too intrusive and not necessarily improve accessibility.

jahow commented on apps/datahub/src/app/record/navigation-bar/navigation-bar.component.html:
```diff
@@ -23,26 +42,42 @@
+  </div>
+  <div
+    class="h-[60px] flex flex-row gap-3 min-w-40 justify-center items-center lg:min-w-60 lg:pl-4 lg:justify-start"
+    [class]="displayMobileMenu ? 'border-l-2 border-gray-300' : ''"
```
Does it work with `[class]`? I thought it needed ngClass:

```suggestion
    [ngClass]="displayMobileMenu ? 'border-l-2 border-gray-300' : ''"
```

tkohr replied:
Yep, double-checked and it works:-)

jahow commented on apps/datahub/src/app/app.module.ts:
```diff
@@ -181,6 +180,7 @@ export const metaReducers: MetaReducer[] = !environment.production ? [] : []
     }),
     OrganisationsComponent,
     LanguageSwitcherComponent,
+    NavigationBarComponent,
```
Is that really needed? this component is part of the same app

tkohr replied:
Ah right, it's not needed, this was a "make it standalone reflex".

jahow commented on apps/datahub/src/app/record/navigation-bar/navigation-bar.component.html:
```diff
@@ -1,20 +1,39 @@
+    class="h-[60px] flex min-w-[60px] items-center justify-center lg:min-w-40 lg:justify-end lg:pr-4"
+    [class]="displayMobileMenu ? 'border-r-2 border-gray-300' : ''"
+  >
+    <gn-ui-navigation-button (click)="back()" [icon]="'matArrowBack'">
```
can't we try using a generic gn-ui-button here? the concept of a "navigation button" is a bit awkward IMO, it's just a button right? plus icons should be properly centered in gn-ui-buttons

tkohr replied:
Right, I'll try to get rid of it.

tkohr replied:
The component has now been removed.

jahow commented on libs/ui/inputs/src/lib/star-toggle/star-toggle.component.html:
```diff
@@ -5,12 +5,15 @@
     }"
   >
-    <ng-icon [name]="toggled ? 'matStar' : 'matStarBorder'"></ng-icon>
+    <ng-icon
+      [name]="toggled ? 'matStar' : 'matStarBorder'"
+      size="24px"
+    ></ng-icon>
```
isn't that going to change the appearance in the search results as well?

tkohr replied:
It's actually the size they are already displayed in, in the search, but for some reason they shrank in the nav bar.

jahow replied:
maybe because the size of the star was proportional to the text size, which is really handy for UI components; fixing a size in px here means it's going to be impossible to have a star of a different size somewhere else

tkohr replied:
Ok I'll see if I get it sized via the text size

jahow commented on libs/ui/layout/src/lib/anchor-link/anchor-link.directive.ts:
```diff
@@ -11,20 +12,29 @@ import {
   @Input('gnUiAnchorLink') targetId: string
   @Input('gnUiAnchorLinkDisabledClass') disabledClass: string
   @Input('gnUiAnchorLinkEnabledClass') enabledClass: string
+  @Input('gnUiAnchorLinkInViewClass') inViewClass: string
```
Remember to include this new input in the associated storybook entry :)

tkohr replied:
Good point, I've added it.

jahow commented on libs/ui/inputs/src/lib/navigation-button/navigation-button.component.ts:
```diff
@@ -6,7 +7,7 @@ import { NgIcon } from '@ng-icons/core'
+  imports: [CommonModule, NgIcon],
   standalone: true,
 })
 export class NavigationButtonComponent {
```
as mentioned above, I think this component should go; everything should use the same buttons

tkohr replied:
It's gone now:-)

jahow commented on libs/ui/inputs/src/lib/star-toggle/star-toggle.component.html:
```diff
@@ -5,12 +5,15 @@
     (click)="toggle($event)"
     [ngClass]="{
       enabled: toggled,
-      'disabled opacity-45': !toggled,
```
this is also going to have impacts on other places right?

tkohr replied:
Yes, do you see another way to make the star black in the nav bar when untoggled to be consistent with the other buttons?

tkohr replied:
Yes it also impacts the display in the results list, which actually fits the new mockups. So I also changed the color from primary to secondary there. Will have to check if this is inline with the current work on the cards.

jahow reviewed (commented):
Thank you, the new functionality of the anchor link directive looks good! I added several components because I feel that layout concerns in the Datahub are leaking all the way through the project, which shouldn't be the case IMO.

tkohr commented:
Thanks for the review @jahow, I'll see what I can do tomorrow.

tkohr commented:
Thanks again for your feedback @jahow ! I've addressed your remarks and rebased the PR on main with the merged [section work](https://github.com/geonetwork/geonetwork-ui/pull/1193). It's now ready for a second review.

LHBruneton-C2C commented:
Thanks for this feature!

I noticed when preparing the demo that the menu is larger than the page content, unlike the mockups.

![image](https://github.com/user-attachments/assets/31619ed8-7b74-4227-a28a-4119b47fbb75)

compared to

![image](https://github.com/user-attachments/assets/cc2c75d3-0402-4ca2-b55e-d475481cbf35)

This should be discussed during the demo, because the default page content width is really smaller than in the mockups.

tkohr commented:
Ah, good point @LHBruneton-C2C, I completely missed that. I'm also noticing that the menu labels are not centered in the mockups.

jahow commented on apps/datahub/src/app/record/navigation-bar/navigation-bar.component.html:
```diff
@@ -1,48 +1,111 @@
+      [attr.data-cy]="l.anchor"
+      gnUiAnchorLink="{{ l.anchor }}"
+      gnUiAnchorLinkDisabledClass="sm:hidden"
+      gnUiAnchorLinkInViewClass="!border-b-primary border-b-4"
```
the bottom padding should be changed here as well, to avoid having the text "jump"

tkohr replied:
Yes, good point, it's fixed now.

jahow commented on apps/datahub/src/app/record/navigation-bar/navigation-bar.component.html:
```diff
@@ -1,48 +1,111 @@
+        [displayCount]="false"
+        class="flex text-background content-center"
+        [style.color]="black"
+        [style.--star-toggle-enabled-color]="black"
+        [style.--star-toggle-disabled-color]="black"
+        [style.--gn-ui-button-rounded]="'8px'"
+        [style.--gn-ui-button-height]="'32px'"
+        [style.--gn-ui-button-width]="'40px'"
+        [style.--gn-ui-button-color]="black"
+        [style.--gn-ui-button-background]="'var(--color-primary-lightest)'"
+        [style.--gn-ui-button-bg-hover]="'var(--color-primary)'"
+        [style.--gn-ui-button-border]="'var(--color-gray-300)'"
+        [style.--gn-ui-button-font-size]="'24px'"
+        [style.--gn-ui-button-padding]="'4px'"
```
FYI for later, css variables can be set on the host in the CSS file :) less verbose in HTML

tkohr commented:
PR is now ready from my side. There was still a mean bug hiding content in the top after using the nav-bar on mobile and scrolling back up. The nav-bar is now really `sticky` instead of `fixed` to solve this.

The latest comments from the PR and the demo have been addressed.

The background and hover colors have been adapted. There was actually a confusion between `bg-primary-lightest` and `bg-primary-white`. Hope it looks better now.

The menu title now updates on mobile when scrolling, as I finally found a solution with little evolution on the directive.

cmoinier reviewed (approved):
Thank you for this cool feature! Works great :)
